docs: dedupe Deep Agents "when to use" and "core capabilities" [closes DOC-1032] (#4022)
langchain-ai/docs: src/oss/deepagents/overview.mdx

Trace requests, debug agent behavior, and evaluate outputs with [LangSmith](https://smith.langchain.com). Follow the [tracing quickstart](/langsmith/trace-with-langchain) to get set up. When ready for production, [deploy to LangSmith Cloud](/langsmith/deploy-to-cloud) for managed hosting.
</Tip>

## Core capabilities

> Use the **Deep Agents SDK** to build agents that handle complex, multi-step tasks across any [model provider](/oss/deepagents/models). The SDK ships with the following built-in capabilities:

Reviewer: ```suggestion
Use the **Deep Agents SDK** to build agents that handle complex, multi-step tasks across **any [model provider](/oss/deepagents/models)**. The SDK ships with the following built-in capabilities:
```

<Card title="Planning and task decomposition" icon="timeline">
    A built-in [`write_todos`](/oss/langchain/middleware/built-in#to-do-list) tool lets agents break down complex tasks into discrete steps, track progress, and adapt plans as new information emerges.
</Card>
<Card title="Context management" icon="scissors">

>     File system tools ([`ls`](/oss/deepagents/harness#virtual-filesystem-access), [`read_file`](/oss/deepagents/harness#virtual-filesystem-access), [`write_file`](/oss/deepagents/harness#virtual-filesystem-access), [`edit_file`](/oss/deepagents/harness#virtual-filesystem-access)) offload large context to storage to prevent context window overflow, and [summarization](/oss/deepagents/context-engineering#summarization) compacts older messages to keep agents effective across extended sessions.

Reviewer 2: it feels slightly wrong to attribute file system tools to offloading context, since this is actually handled via middleware business logic. it is correct that context lands in the filesystem

also, summarization offloads in addition to compacting. not sure if we need to make that distinction

</Card>

> <Card title="Pluggable filesystem backends" icon="plug">
>     Swap the virtual filesystem to fit your use case via [pluggable backends](/oss/deepagents/backends): in-memory state, local disk, LangGraph store for cross-thread persistence, [sandboxes](/oss/deepagents/sandboxes) (Modal, Daytona, Deno), composite routing, or your own custom backend.
> </Card>

Reviewer: ```suggestion
<Card title="Pluggable filesystem backends" icon="plug">
    Swap the virtual filesystem to fit your use case via [pluggable backends](/oss/deepagents/backends): in-memory state, local disk, LangGraph store for cross-thread persistence, [sandboxes](/oss/deepagents/sandboxes) (Modal, Daytona, Deno), composite routing, or your own custom backend with declarative [permission rules](/oss/deepagents/permissions) that restrict which files agents can read or write.
</Card>
```

>     Swap the virtual filesystem to fit your use case via [pluggable backends](/oss/deepagents/backends): in-memory state, local disk, LangGraph store for cross-thread persistence, [sandboxes](/oss/deepagents/sandboxes) (Modal, Daytona, Deno), composite routing, or your own custom backend with declarative [permission rules](/oss/deepagents/permissions) that restrict which files agents can read or write.

Reviewer 2: not sure sandboxes belong here

<Card title="Shell execution" icon="terminal">

>     With a [sandbox backend](/oss/deepagents/sandboxes), agents get an `execute` tool to run shell commands for tests, builds, git operations, and system tasks — isolated from your host system.

Reviewer 2: this frames it like you need to use a remote sandbox to get the `execute` tool, which is not true since `LocalShellBackend` provides it

it is true that for real isolation you need a remote sandbox though

</Card>
<Card title="Interpreters" icon="code">
    Add an [interpreter](/oss/deepagents/interpreters) to run JavaScript in an in-memory runtime. Interpreters let agents compose tools programmatically, orchestrate subagents, and transform structured data without a full shell environment.
</Card>
<Card title="Subagent spawning" icon="users-group">

>     A built-in `task` tool spawns specialized subagents for context isolation, keeping the main agent's context clean while going deep on specific subtasks.

Reviewer: ```suggestion
    A built-in `task` tool spawns general-purpose or specialized subagents for context isolation, keeping the main agent's context clean while going deep on specific subtasks.
```

>     A built-in `task` tool spawns general-purpose or specialized subagents for context isolation, keeping the main agent's context clean while going deep on specific subtasks.

Reviewer 2: should we mention async subagent option?

</Card>
<Card title="Long-term memory" icon="database">
    Persist memory across threads and conversations using LangGraph's [Memory Store](/oss/langgraph/persistence#memory-store).
</Card>
<Card title="Filesystem permissions" icon="lock">
    Declare [permission rules](/oss/deepagents/permissions) that control which files and directories agents can read or write. Subagents can inherit or override the parent's rules.
</Card>
<Card title="Human-in-the-loop" icon="user-check">
    Configure [human approval](/oss/deepagents/human-in-the-loop) for sensitive tool operations using LangGraph's interrupt capabilities.
</Card>
<Card title="Skills" icon="puzzle">
    Extend agents with reusable [skills](/oss/deepagents/skills) that provide specialized workflows, domain knowledge, and custom instructions.
</Card>
<Card title="Smart defaults" icon="wand">
    Ships with opinionated system prompts that teach the model to plan before acting, verify work, and manage context. Customize or replace the defaults as needed.
</Card>

> :::python
> For building simpler agents, consider using LangChain's [`create_agent`](/oss/langchain/agents) or building a custom [LangGraph](/oss/langgraph/overview) workflow.
> :::
> 
> :::js
> For building simpler agents, consider using LangChain's [`createAgent`](/oss/langchain/agents) or building a custom [LangGraph](/oss/langgraph/overview) workflow.
> :::

Reviewer: ```suggestion
:::python
For building custom agents without these builtin capabilities, consider using LangChain's [`create_agent`](/oss/langchain/agents) or building a custom [LangGraph](/oss/langgraph/overview) workflow.
:::

:::js
For building custom agents without these builtin capabilities, consider using LangChain's [`createAgent`](/oss/langchain/agents) or building a custom [LangGraph](/oss/langgraph/overview) workflow.
:::
```
